soroban-env-host/src/native_contract/token/balance.rs

```diff
@@ -6,74 +6,150 @@ use crate::native_contract::token::storage_types::DataKey;
 use crate::{err, HostError};
 use core::cmp::Ordering;
 use soroban_env_common::xdr::{
-    AccountEntryExt, AccountEntryExtensionV1Ext, AccountId, LedgerEntryData, TrustLineAsset,
-    TrustLineEntryExt, TrustLineFlags,
+    AccountEntryExt, AccountEntryExtensionV1Ext, AccountFlags, AccountId, LedgerEntryData,
+    TrustLineAsset, TrustLineEntryExt, TrustLineFlags,
 };
 use soroban_env_common::{CheckedEnv, TryIntoVal};
 
 use super::error::ContractError;
+use super::storage_types::BalanceValue;
 
 // Metering: *mostly* covered by components. Not sure about `try_into_val`.
 pub fn read_balance(e: &Host, id: Identifier) -> Result<BigInt, HostError> {
     let key = DataKey::Balance(id);
     if let Ok(balance) = e.get_contract_data(key.try_into_val(e)?) {
-        Ok(balance.try_into_val(e)?)
+        let balance: BalanceValue = balance.try_into_val(e)?;
+        Ok(balance.amount)
     } else {
         Ok(BigInt::from_u64(e, 0)?)
     }
 }
 
+fn is_issuer_auth_required(e: &Host, issuer_id: AccountId) -> Result<bool, HostError> {
+    let lk = e.to_account_key(issuer_id.clone());
+    e.with_mut_storage(|storage| {
+        //TODO: Should we add a try_get method to storage?
+        if storage.has(&lk)? {
+            let le = storage.get(&lk)?;
+            let auth_required = match le.data {
+                LedgerEntryData::Account(ae) => {
+                    (ae.flags & (AccountFlags::RequiredFlag as u32)) != 0
+                }
+                _ => false, //TODO: This shouldn't happen... should we error instead?
```

Comment:
Normally we return the internal contract error in such cases (panic could be valid, but it's a bit risky).

Reply:
Updated to return an internal contract error.